### ISSUE TITLE
SORT in spliced subqueries sometimes forgets documents

### ISSUE THREAD

**Issue by wajda** [arangodb/arangodb#12693]
## My Environment

* __ArangoDB Version__:        3.7.2-1
* __Storage Engine__:             RocksDB
* __Deployment Mode__:         Single Server
* __Deployment Strategy__:     Manual Start
* __Configuration__:               all by default
* __Infrastructure__:               local PC
* __Operating System__:        Linux, Fedora 32
* __Total RAM in your machine__:        32Gb
* __Disks in use__:        SSD
* __Used Package__:              rpm

## Component, Query & Data

__Affected feature__:
AQL query

__AQL query (if applicable)__:
```
FOR ep IN executionPlan
    SORT ep._id
    LIMIT 100, 400
    
    // traverse the graph and collect names of vesited vertex collections
    LET names = (
        FOR v IN 1..999
            OUTBOUND ep executes, follows, affects
            COLLECT name = PARSE_IDENTIFIER(v._id).collection INTO g
            RETURN name
    )

    // execute two identical grouping queries on the 'names' array and store the results in two variables 'xs1' and 'xs2'
    LET xs1 = (FOR n IN names COLLECT c = n INTO g RETURN c)
    LET xs2 = (FOR n IN names COLLECT c = n INTO g RETURN c)

    // 'xs1' and 'xs2' should be equal as they are produced by identical expressions,
    // thus I expect this filter to never pass and the result of the entire AQL query to be empty.
    FILTER xs1 != xs2
    
    // Unexpectedly enabling any of the filters below magically changes the content of 'xs1',
    // making the above filter pass!
    
    ///////////////////////////////////////////////////////////////////
    FILTER NOT POSITION(xs1, 'operation')
    // FILTER LENGTH(xs1[* FILTER CURRENT == 'operation']) == 0
    ///////////////////////////////////////////////////////////////////
    
    // Normally the result should be empty because 'xs1 == xs2',
    // but enabling one of the filters from a box above emits values. 
    RETURN {names, xs1, xs2}
```
__AQL explain and/or profile (if applicable)__:
```
Query String (1343 chars, cacheable: true):
 FOR ep IN executionPlan
     SORT ep._id
     LIMIT 100, 400
     
     // traverse the graph and collect names of vesited vertex collections
     LET names = (
         FOR v IN 1..999
             OUTBOUND ep executes, follows, affects
             COLLECT name = PARSE_IDENTIFIER(v._id).collection INTO g
             RETURN name
     )
     // execute two identical grouping queries on the 'names' array and store the results in two 
 variables 'xs1' and 'xs2'
     LET xs1 = (FOR n IN names COLLECT c = n INTO g RETURN c)
     LET xs2 = (FOR n IN names COLLECT c = n INTO g RETURN c)
     // 'xs1' and 'xs2' should be equal as they are produced by identical expressions,
     // thus I expect this filter to never pass and the result of the entire AQL query to be empty.
     FILTER xs1 != xs2
     
     // Unexpectedly enabling any of the filters below magically changes the content of 'xs1',
     // making the above filter pass!
     
     ///////////////////////////////////////////////////////////////////
     FI...

Execution plan:
 Id   NodeType                           Est.   Comment
  1   SingletonNode                         1   * ROOT
  2   EnumerateCollectionNode          111918     - FOR ep IN executionPlan   /* full collection scan */
  3   CalculationNode                  111918       - LET #19 = ep.`_id`   /* attribute expression */   /* collections used: ep : executionPlan */
  4   SortNode                         111918       - SORT #19 ASC   /* sorting strategy: constrained heap */
  5   LimitNode                           400       - LIMIT 100, 400
 35   SubqueryStartNode                   400       - LET names = ( /* subquery begin */
  7   TraversalNode                 583994610         - FOR v  /* vertex */ IN 1..999  /* min..maxPathDepth */ OUTBOUND ep /* startnode */  executes, OUTBOUND follows, OUTBOUND affects
  8   CalculationNode               583994610           - LET #23 = PARSE_IDENTIFIER(v.`_id`).`collection`   /* simple expression */
 28   SortNode                      583994610           - SORT #23 ASC   /* sorting strategy: standard */
  9   CollectNode                   467195688           - COLLECT name = #23   /* sorted */
 36   SubqueryEndNode               467195688           - RETURN  name ) /* subquery end */
 33   SubqueryStartNode             467195688       - LET xs1 = ( /* subquery begin */
 13   EnumerateListNode           46719568800         - FOR n IN names   /* list iteration */
 29   SortNode                    46719568800           - SORT n ASC   /* sorting strategy: standard */
 14   CollectNode                 37375655040           - COLLECT c = n   /* sorted */
 34   SubqueryEndNode             37375655040           - RETURN  c ) /* subquery end */
 24   CalculationNode             37375655040       - LET #27 = ! POSITION(xs1, "operation")   /* simple expression */
 25   FilterNode                  37375655040       - FILTER #27
 31   SubqueryStartNode           37375655040       - LET xs2 = ( /* subquery begin */
 18   EnumerateListNode         3737565504000         - FOR n IN names   /* list iteration */
 30   SortNode                  3737565504000           - SORT n ASC   /* sorting strategy: standard */
 19   CollectNode               2990052403200           - COLLECT c = n   /* sorted */
 32   SubqueryEndNode           2990052403200           - RETURN  c ) /* subquery end */
 22   CalculationNode           2990052403200       - LET #25 = (xs1 != xs2)   /* simple expression */
 23   FilterNode                2990052403200       - FILTER #25
 26   CalculationNode           2990052403200       - LET #29 = { "names" : names, "xs1" : xs1, "xs2" : xs2 }   /* simple expression */
 27   ReturnNode                2990052403200       - RETURN #29

Indexes used:
 By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields        Ranges
  7   edge   edge   affects      false    false        99.99 %   [ `_from` ]   base OUTBOUND
  7   edge   edge   executes     false    false        99.98 %   [ `_from` ]   base OUTBOUND
  7   edge   edge   follows      false    false        95.48 %   [ `_from` ]   base OUTBOUND

Functions used:
 Name               Deterministic   Cacheable   Uses V8
 PARSE_IDENTIFIER   true            true        false  
 POSITION           true            true        false  

Traversals on graphs:
 Id  Depth   Vertex collections  Edge collections            Options                                  Filter / Prune Conditions
 7   1..999                      executes, follows, affects  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   remove-collect-variables
  6   sort-limit
  7   splice-subqueries

```

__Dataset__:
Due to unpredictable behavior of the query I was not able to reproduce it on a small enough database that I can share here.
Will probably add it later.  

__Size of your Dataset on disk__:
around 5.5Gb

__Replication Factor & Number of Shards (Cluster only)__:
single node setup. No replicas, no shards.

## Steps to reproduce

1.  Comment the outlined `FILTER` statements.
2.  Verify the content of variables `xs1` and `xs2` is the same, and the query returns empty result as expected.    
3.  Uncomment any of the outlined `FILTER` statements (e.g. `FILTER NOT POSITION(xs1, 'operation')`)   
4.  Variable `xs1` gets changed and no longer equal to `xs2`. The query returns the record, proving the discrepancy.

__Problem__:
The above AQL FILTER statement somehow changes the content of the variable they examine.
Logically (assuming no side effect inside AQL) the filter statement following `FILTER x1 != x2` should have no effect on the query, because `xs1 == xs2` by definition (they are produced by identical expressions), so all records should be filtered out and the result of the query should be empty.

But somehow adding, for example `FILTER NOT POSITION(xs1, 'operation')` (no matter with or without `NOT`) changes the value of `xs1` making two variable different, passing the first filter and result to be returned.

Another strange thing is that this behavior isn't consistent in different variations of the query. Making logically insignificant changes like removing unused `INTO g` statement or redundant `COLLECT c=n` makes the strange behavior to disappear.

Yet another strange thing - replacing `LIMIT 500` with `LIMIT 100, 400` makes the query to return more records (2 instead of 1)

__Expected result__:
Empty result

**Comment by wajda** [comment]
![1](https://user-images.githubusercontent.com/795479/94034346-32717300-fdc2-11ea-8518-a4303ffe4352.png)

**Comment by wajda** [comment]
![2](https://user-images.githubusercontent.com/795479/94034361-3a311780-fdc2-11ea-8b17-84ef90edefde.png)

**Comment by goedderz** [comment]
Hi @wajda,

thanks for your report! Would you be so kind and provide the output `explainRegisters` of the broken version of the query? Enter the following in `arangosh`:
```
> var query = `FOR ep IN ...
...
`
> require('@arangodb/aql/explainer').explainRegisters(query)
```
And of course copy the whole query between the backticks \`\`.

Best,

Tobias

**Comment by wajda** [comment]
Hm, I don't seem to have `explainRegisters` function. Only `explain` is there.
```shell
127.0.0.1:8529@spline-jedi> require('@arangodb/aql/explainer')
{ 
  "explain" : [Function "explain(data, options, shouldPrin" ...], 
  "profileQuery" : [Function "profileQuery(data, shouldPrin" ...], 
  "debug" : [Function "debug(query, bindVars, option" ...], 
  "debugDump" : [Function "debugDump(filename, query, bindVars, option" ...], 
  "inspectDump" : [Function "inspectDump(filename, outfil" ...] 
}
```

**Comment by wajda** [comment]
Arangosh version I'm using:
```
arangosh (ArangoDB 3.7.2-1 [linux] 64bit, using jemalloc, build tags/v3.7.2.1-0-g2c56dec4b2c-dirty, VPack 0.1.33, RocksDB 6.8.0, ICU 64.2, V8 7.9.317, OpenSSL 1.1.1g  21 Apr 2020)
```

**Comment by jsteemann** [comment]
@goedderz @wajda : to clarify, 3.7 doesn't have `explainRegisters` yet. This is only present in 3.8/devel.

**Comment by wajda** [comment]
Replicated it in a nightly build of 3.8 in a docker.

```
127.0.0.1:8529@spline-jedi> require('@arangodb/aql/explainer').explainRegisters(query)
2020-09-23T22:57:23Z [488] WARNING [ebe22] {general} explainRegisters() is purposefully undocumented and may be changed or removed without notice.
 id│         type          │depth 
───┼───────────────────────┼──────
  1│SingletonNode          │    0   ││
   │                       │        ││
  2│EnumerateCollectionNode│    1   │   →ep    │     │
   │                       │        │          │     │
  3│CalculationNode        │    1   │   ←ep    │→#19 │
   │                       │        │    ↓     │     │
  4│SortNode               │    1   │    ep    │←#19 │
   │                       │        │    ↓     │     │
  5│LimitNode              │    1   │    ep    │  -  │
   │                       │        │    ↓     │     │
 35│SubqueryStartNode      │    2   │    ep    │  -  │
   │                       │        │          │     │
↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘  │          │     │
 35│SubqueryStartNode      │    2   │    ep    │  -  │
   │                       │        │    ↓     │     │
  7│TraversalNode          │    3   │ ←ep →v   │  -  │
   │                       │        │          │     │
  8│CalculationNode        │    3   │   ←v     │→#23 │
   │                       │        │    ⮾     │     │
 28│SortNode               │    3   │    -     │←#23 │
   │                       │        │          │  ↓  │
  9│CollectNode            │    4   │  →name   │←#23 │
   │                       │        │          │     │
 36│SubqueryEndNode        │    5   │  ←name   │  -  │
↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙  │          │     │
 36│SubqueryEndNode        │    5   │ →names   │     │
   │                       │        │          │     │
 33│SubqueryStartNode      │    6   │  names   │  -  │
   │                       │        │    ↓     │     │
↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘  │          │     │
 33│SubqueryStartNode      │    6   │  names   │  -  │
   │                       │        │    ↓     │     │
 13│EnumerateListNode      │    7   │←names →n │  -  │
   │                       │        │          │     │
 29│SortNode               │    7   │   ←n     │  -  │
   │                       │        │    ↓     │     │
 14│CollectNode            │    8   │  ←n →c   │  -  │
   │                       │        │          │     │
 34│SubqueryEndNode        │    9   │   ←c     │  -  │
↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙  │          │     │
 34│SubqueryEndNode        │    9   │  names   │→xs1 │     │
   │                       │        │    ↓     │     │     │
 24│CalculationNode        │    9   │  names   │←xs1 │→#27 │
   │                       │        │    ↓     │  ↓  │     │
 25│FilterNode             │    9   │  names   │ xs1 │←#27 │
   │                       │        │    ↓     │  ↓  │     │
 31│SubqueryStartNode      │   10   │  names   │ xs1 │  -  │
   │                       │        │    ↓     │  ↓  │     │
↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘↘  │          │     │     │
 31│SubqueryStartNode      │   10   │  names   │ xs1 │  -  │
   │                       │        │    ↓     │     │     │
 18│EnumerateListNode      │   11   │←names →n │  -  │  -  │
   │                       │        │          │     │     │
 30│SortNode               │   11   │   ←n     │  -  │  -  │
   │                       │        │    ↓     │     │     │
 19│CollectNode            │   12   │  ←n →c   │  -  │  -  │
   │                       │        │          │     │     │
 32│SubqueryEndNode        │   13   │   ←c     │  -  │  -  │
↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙↙  │          │     │     │
 32│SubqueryEndNode        │   13   │  names   │ xs1 │→xs2 │     │
   │                       │        │    ↓     │  ↓  │     │     │
 22│CalculationNode        │   13   │  names   │←xs1 │←xs2 │→#25 │
   │                       │        │    ↓     │  ↓  │  ↓  │     │
 23│FilterNode             │   13   │  names   │ xs1 │ xs2 │←#25 │
   │                       │        │    ↓     │  ↓  │  ↓  │     │
 26│CalculationNode        │   13   │ ←names   │←xs1 │←xs2 │→#29 │
   │                       │        │    ⮾     │  ⮾  │  ⮾  │     │
 27│ReturnNode             │   13   │    -     │  -  │  -  │←#29 │
   │                       │        │          │     │     │     │

 Legend:
 [ ⮾ ]: Clear register
 [ ↓ ]: Keep register
 [ → ]: Write register
 [ ← ]: Read register
 [ - ]: Register is unused
 [ ↘ ]: Entering a subquery
 [ ↙ ]: Leaving a subquery


127.0.0.1:8529@spline-jedi> 
```

**Comment by goedderz** [comment]
> @goedderz @wajda : to clarify, 3.7 doesn't have `explainRegisters` yet. This is only present in 3.8/devel.

Ah, I remembered that wrong, thanks.

> Replicated it in a nightly build of 3.8 in a docker.

Thanks @wajda! I will try to reproduce the problem.

**Comment by wajda** [comment]
I was able to cut my database to a size suitable for uploading it here for you to reproduce the issue easily. (see attachment)

```shell
$ arangorestore --input-directory arangodb.issue.12693.dump --server.endpoint tcp://localhost:8529 --server.database github-12693 --create-database
```

AQL that reproduces the issue:
```aql
FOR ep IN executionPlan
    LET names = (
        FOR v IN 1..999
            OUTBOUND ep executes, follows, affects
            COLLECT name = PARSE_IDENTIFIER(v._id).collection INTO g
            RETURN name
    )

    LET xs1 = (FOR n IN names COLLECT c = n INTO g RETURN c)
    LET xs2 = (FOR n IN names COLLECT c = n INTO g RETURN c)

    FILTER xs1 != xs2
    
    FILTER NOT POSITION(xs1, 'operation')
    
    RETURN {names, xs1, xs2}
``` 

Expected behavior - empty result.
Actual behavior - an object with different `xs1` and `xs2`

[arangodb.issue.12693.dump.tar.gz](https://github.com/arangodb/arangodb/files/5299607/arangodb.issue.12693.dump.tar.gz)

**Comment by goedderz** [comment]
Thanks @wajda, that will help a lot!

**Comment by goedderz** [comment]
FYI, I can reproduce the problem.

**Comment by wajda** [comment]
Does the above PR fix the issue or is it still work in progress? 
Is there any nightly build with the fix to test?

**Comment by goedderz** [comment]
It does fix the issue as far as I can tell. It is still work in progress, because there are some conflicts with existing unit tests that need to be resolved. I am working on that. After the PR is merged it will be available in the next _devel_ nightly build. I will make a backport PR for 3.7 as soon as this PR is finished.

**Comment by goedderz** [comment]
Hi @wajda,

the fix for 3.7 is merged and will be available in 3.7.3. It might not have made it in the current 3.7 nightly, but will be part of the next one.

Best,

Tobias

**Comment by wajda** [comment]
Seems to be working so far. Thank you very much!
Will test it more on a bigger db and will let you know if the issue returns.

**Comment by goedderz** [comment]
Glad to hear that, and thanks for verifying!

Fyi, and sorry I didn't mention this earlier: [disabling the optimizer rule](https://www.arangodb.com/docs/stable/aql/execution-and-performance-optimizer.html#turning-specific-optimizer-rules-off) `splice-subqueries` should be a viable workaround, until the fix is released.

**Comment by maxkernbach** [comment]
Hi @wajda,

Version 3.7.3, which includes a fix for this issue, has just been released.

From the [Changelog](https://github.com/arangodb/arangodb/blob/19b71a404cc119c65589f8da0c36e3a70e1eddd4/CHANGELOG#L77):

> * Fix #12693: SORT inside a subquery could sometimes swallow part of its input
 when it crossed boundaries of internal row batches.

Thanks again for reporting. In case this problem still occurs, please comment.
